confmap/confmap.go

```diff
@@ -273,20 +273,25 @@ func unmarshalerHookFunc(result interface{}) mapstructure.DecodeHookFuncValue {
 		}
 
 		toPtr := to.Addr().Interface()
-		if _, ok := toPtr.(Unmarshaler); !ok {
+		// Need to ignore the top structure to avoid circular dependency.
+		if toPtr == result {
```

Comment:
Is this moved up to avoid an unnecessary allocation for `unmarshaler`?

Reply:
You caught me :)) Now, I am blocked by https://github.com/open-telemetry/opentelemetry-collector/pull/6392